### PR TITLE
Allow multiple observers for the Navigator and MaterialApp

### LINKED ISSUE #7831: Allow multiple observers for the navigator
We would like make the [Navigator](https://docs.flutter.io/flutter/widgets/Navigator-class.html) accept a list of [observers](https://docs.flutter.io/flutter/widgets/NavigatorObserver-class.html). So that we can track page translation automatically using Google Analytics. Currently, in MateralApp, the observer is only used for [managing Hero transitions ](https://docs.flutter.io/flutter/widgets/HeroController-class.html).

Since navigator is usually generated from MaterialApp or WidgetsApp, we need to modify these classes to accept a list of observers as well.

### PR DESCRIPTION
**Pull request by 123qws**
Fix #7831

### CODE REVIEW

File: packages/flutter/test/widgets/navigator_test.dart [hunk 1]
```diff
@@ -260,6 +286,41 @@ void main() {
...
+       '/': (BuildContext context) => new OnTapPage(id: '/', onTap: () { Navigator.pushNamed(context, '/A'); }),
+      '/A': (BuildContext context) => new OnTapPage(id: 'A', onTap: () { Navigator.pop(context); }),
+    };
+    TestObserver observer1 = new TestObserver();
+    TestObserver observer2 = new TestObserver();
+
+    await tester.pumpWidget(new MaterialApp(
+      routes: routes,
+      navigatorObservers: <NavigatorObserver>[observer1, observer2],
+    ));
```
Comment (abarth) [inline packages/flutter/test/widgets/navigator_test.dart]: Can you test that adding and removing an observer works as expected?

**Comment by abarth** [comment]
style nit: loops that have a single statement should omit `{` and `}` (see https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-using-braces-for-one-line-long-control-structure-statements )

File: packages/flutter/lib/src/material/app.dart [hunk 2]
```diff
@@ -154,6 +155,9 @@ class MaterialApp extends StatefulWidget {
   /// representative of what will happen in release mode.
   final bool debugShowCheckedModeBanner;
 
+  /// The list of observers for the Navigator created for this app.
```
Comment (abarth) [inline packages/flutter/lib/src/material/app.dart]: `s/Navigator/[Navigator]/`

Using the `[` braces will make the name a link in the docs web site.

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 3]
```diff
@@ -678,18 +680,22 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
 
   @override
   void didUpdateConfig(Navigator oldConfig) {
-    if (oldConfig.observer != config.observer) {
-      oldConfig.observer?._navigator = null;
-      assert(config.observer == null || config.observer.navigator == null);
-      config.observer?._navigator = this;
+    for (NavigatorObserver observer in oldConfig.observers) {
+      observer._navigator = null;
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/navigator.dart]: nit: no braces on one-line for blocks

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 4]
```diff
@@ -678,18 +680,22 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
...
+      observer._navigator = this;
     }
   }
 
   @override
   void dispose() {
     assert(!_debugLocked);
     assert(() { _debugLocked = true; return true; });
-    config.observer?._navigator = null;
+    for (NavigatorObserver observer in config.observers) {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/navigator.dart]: nit: no braces on one-line for blocks

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 5]
```diff
@@ -768,7 +774,9 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
       route.didChangeNext(null);
       if (oldRoute != null)
         oldRoute.didChangeNext(route);
-      config.observer?.didPush(route, oldRoute);
+      for (NavigatorObserver observer in config.observers) {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/navigator.dart]: nit: no braces on one-line for blocks

File: packages/flutter/lib/src/widgets/app.dart [hunk 6]
```diff
@@ -111,8 +111,8 @@ class WidgetsApp extends StatefulWidget {
   /// representative of what will happen in release mode.
   final bool debugShowCheckedModeBanner;
 
-  /// The observer for the Navigator created for this app.
-  final NavigatorObserver navigatorObserver;
+  /// The list of observers for the Navigator created for this app.
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/app.dart]: `s/Navigator/[Navigator]/`

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 7]
```diff
@@ -850,7 +858,9 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
       newRoute.didChangeNext(null);
       if (index > 0)
         _history[index - 1].didChangeNext(newRoute);
-      config.observer?.didPush(newRoute, oldRoute);
+      for (NavigatorObserver observer in config.observers) {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/navigator.dart]: nit: no braces on one-line for blocks

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 8]
```diff
@@ -957,7 +967,9 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
           if (route._navigator != null)
             _poppedRoutes.add(route);
           _history.last.didPopNext(route);
-          config.observer?.didPop(route, _history.last);
+          for (NavigatorObserver observer in config.observers) {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/navigator.dart]: nit: no braces on one-line for blocks

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 9]
```diff
@@ -1026,13 +1038,17 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
   /// Used for the iOS back gesture.
   void didStartUserGesture() {
     _userGestureInProgress = true;
-    config.observer?.didStartUserGesture();
+    for (NavigatorObserver observer in config.observers) {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/navigator.dart]: nit: no braces on one-line for blocks

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 10]
```diff
@@ -678,18 +680,22 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
 
   @override
   void didUpdateConfig(Navigator oldConfig) {
-    if (oldConfig.observer != config.observer) {
```
Comment (abarth) [inline packages/flutter/lib/src/widgets/navigator.dart]: I'd keep this equality check.  That lets a client who has a really complex set of observers avoid the loops if they do the work of re-using the same list when it doesn't change.

File: packages/flutter/lib/src/widgets/navigator.dart [hunk 11]
```diff
@@ -1026,13 +1038,17 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
...
+    for (NavigatorObserver observer in config.observers) {
+      observer.didStartUserGesture();
+    }
   }
 
   /// A user gesture is no longer controlling the navigator.
   void didStopUserGesture() {
     _userGestureInProgress = false;
-    config.observer?.didStopUserGesture();
+    for (NavigatorObserver observer in config.observers) {
```
Comment (Hixie) [inline packages/flutter/lib/src/widgets/navigator.dart]: nit: no braces on one-line for blocks

**Comment by abarth** [comment]
LGTM w/ a test for adding/removing observers.

File: packages/flutter/test/widgets/navigator_test.dart (hunk 1 above)
Comment (weisong0) [inline packages/flutter/test/widgets/navigator_test.dart]: Done.

File: packages/flutter/lib/src/widgets/navigator.dart (hunk 10 above)
Comment (weisong0) [inline packages/flutter/lib/src/widgets/navigator.dart]: Like it!

**Comment by weisong0** [comment]
Thanks for your reviews! PTAL.

**Comment by Hixie** [comment]
LGTM

Please let us know if you need someone to check this in.

**Comment by weisong0** [comment]
Can you help me check this in? Thanks!

**Pull request merged**